.travis.yml

```diff
@@ -89,45 +114,70 @@ install:
       ${TRAVIS_BUILD_DIR}/contrib/build-pybind11-in-conda.sh
       # Debugging information
       conda info -a
-    else
+    elif [[ "$RUNMODE" == "docker"  ]]; then
       $SCRIPT_RUN_PREFIX id
       $SCRIPT_RUN_PREFIX apt-get -qq update
       $SCRIPT_RUN_PREFIX apt-get -qqy install fakeroot debhelper locales \
         libreadline7 libssl1.0.0 libffi6 \
         liblapack3 liblapack-dev libhdf5-100 libhdf5-dev libnetcdf13 \
         libnetcdf-dev libscotch-6.0 libscotch-dev cython3 python3 \
         python3-numpy libpython3.6-dev python3-boto python3-paramiko graphviz
+    elif [[ "$RUNMODE" == "singularity"  ]]; then
+      pushd ${SHOME}
+      cp ${TRAVIS_BUILD_DIR}/sandbox/gas/tube/${SOLVCON_DRIVING} ${SHOME}
+      # trick to inject TRAVIS_BRANCH and TRAVIS_REPO_SLUG
+      echo '#!/bin/bash' > ${SHOME}/prepare-solvcon-dev.sh
+      echo "export TRAVIS_BRANCH=${TRAVIS_BRANCH}" >> ${SHOME}/prepare-solvcon-dev.sh
+      echo "export TRAVIS_REPO_SLUG=${TRAVIS_REPO_SLUG}" >> ${SHOME}/prepare-solvcon-dev.sh
+      cat ${TRAVIS_BUILD_DIR}/contrib/singularity/prepare-solvcon-dev.sh >> ${SHOME}/prepare-solvcon-dev.sh
+      cp ${TRAVIS_BUILD_DIR}/contrib/singularity/${SRECIPE} ${SHOME}
+      sudo singularity build ${SIMAGE} ${SRECIPE}
+      ls
+      popd
     fi
 
 script:
   # Debugging information
   - |
-    $SCRIPT_RUN_PREFIX which python3
-    $SCRIPT_RUN_PREFIX python3 --version
-    $SCRIPT_RUN_PREFIX which $CXX
-    $SCRIPT_RUN_PREFIX $CXX --version
-    $SCRIPT_RUN_PREFIX which $CC
-    $SCRIPT_RUN_PREFIX $CC --version
-  - $SCRIPT_RUN_PREFIX make SC_PURE_PYTHON=1 test_from_package
+    # We do everything in the singularity container including unit tests. The
+    # external environment could be very simple and does not have python3, so
+    # we do not need to dump the following debugging information otherwise it
+    # causes none-zero exit return code.
+    if [[ "$RUNMODE" != "singularity" ]] ; then
+      $SCRIPT_RUN_PREFIX which python3
+      $SCRIPT_RUN_PREFIX python3 --version
+      $SCRIPT_RUN_PREFIX which $CXX
+      $SCRIPT_RUN_PREFIX $CXX --version
+      $SCRIPT_RUN_PREFIX which $CC
+      $SCRIPT_RUN_PREFIX $CC --version
+    fi
+  - |
+    if [[ "$RUNMODE" != "singularity" ]] ; then
+      $SCRIPT_RUN_PREFIX make SC_PURE_PYTHON=1 test_from_package
+    fi
   - |
     if [[ "$RUNMODE" == "docker" ]] ; then
       $SCRIPT_RUN_PREFIX make deb
       $SCRIPT_RUN_PREFIX dpkg -i dist/debbuild/*.deb
       $SCRIPT_RUN_PREFIX bash -c "cd /tmp ; python3 -c 'import solvcon ; print(solvcon)'"
       $SCRIPT_RUN_PREFIX bash -c "cd /tmp ; python3 -c 'import solvcon ; solvcon.test()'"
+    elif [[ "$RUNMODE" == "singularity" ]] ; then
+      $SCRIPT_RUN_PREFIX ${TRAVIS_BUILD_DIR}/sandbox/gas/tube/${SOLVCON_DRIVING} run tube
     fi
 
 after_script:
   - |
     if [[ "$RUNMODE" == "docker" ]] ; then
       docker stop "$containerid"
       docker rm "$containerid"
+    elif [[ "$RUNMODE" == "singularity" ]] ; then
```

Comment:
This is another reason that I don't like to build singularity from source right here.